libsnark/zk_proof_systems/plonk/srs.tcc

```diff
@@ -106,12 +106,128 @@ usrs<ppT> plonk_usrs_derive_from_secret(
+    // unity and a generator of H i.e: H = {1, w, ... , w^{n-1}}. We assume
+    // that the number of gates in a circuit is no more than n. We also include
+    // an optimisation suggested by Vitalik Buterin, to define the identity per-
+    // mutations through degree-1 polynomials. The identity permutations must
+    // map each wire value to a unique element \in F. This can be done by
+    // defining S_ID1(X) = X, S_ID2 (X) = k1 X, S_ID3(X) = k2 X [see below for
+    // more on S_ID1, S_ID2, S_ID3], where k1 , k2 are quadratic non-residues
+    // \in F. This effectively maps each wire value to a root of unity in H,
+    // with right and output wires having an additional multiplicative factor of
+    // k1, k2 applied respectively. By representing the identity permutation via
+    // degree-1 polynomials, their evaluations can be directly computed by the
+    // verifier. This reduces the size of the proof by 1 F element, as well as
+    // reducing the number of Fast-Fourier-Transforms required by the prover."
+    //
+    // Further in Sect. 8.1 [GWC19]:
+    //
+    // "S_ID1(X) = X, S_ID2(X) = k 1 X, S ID3 (X) = k 2 X: the identity
+    // permutation applied to a, b, c [the wire polynomials, see Round 1, p.27
+    // [GWC19]]. k1, k2 \in F are chosen such that H, k1 H, k2 H are distinct
+    // cosets of H in F*, and thus consist of 3n distinct elements. (For
+    // example, when w (omega) is a quadratic residue in F, take k1 to be any
+    // quadratic non-residue, and k2 to be a quadratic non-residue not
+    // contained in k1 H.)"
+    //
+    // For the moment k1,k2 are fixed to the test vector values for debug
+    // purpouses. TODO choose k1,k2 according to the requirements in [GWC19]
+    libff::Fr<ppT> k1 =
+        Field("706987411474581393682955260879121390206111740035659671471"
+              "3673571023200548519");
+    libff::Fr<ppT> k2 = libff::power(k1, libff::bigint<1>(2));
+#ifdef DEBUG_PLONK
+    printf("[%s:%d] k1 ", __FILE__, __LINE__);
+    k1.print();
+    printf("[%s:%d] k2 ", __FILE__, __LINE__);
+    k2.print();
+#endif // #ifdef DEBUG_PLONK
+
+    // omega[0] are the n roots of unity; omega[1] are omega[0]*k1;
+    // omega[2] are omega[0]*k2
+    std::vector<std::vector<Field>> omega_roots;
+    plonk_compute_roots_of_unity_omega(num_gates, k1, k2, omega_roots);
+
+    // H_gen contains the generators of H, k1 H and k2 H in one place
+    // ie. circuit.omega_roots, circuit.omega_roots_k1 and
+    // circuit.omega_roots_k2
+    std::vector<Field> H_gen;
+    plonk_compute_cosets_H_k1H_k2H(num_gates, k1, k2, H_gen);
+
+    // TODO: write unit test for plonk_roots_of_unity_omega_to_subgroup_H
+#ifdef DEBUG_PLONK
+    printf("[%s:%d] H_gen\n", __FILE__, __LINE__);
+    libff::print_vector(H_gen);
+    for (int i = 0; i < (int)H_gen.size(); ++i) {
+        assert(H_gen[i] == example.H_gen[i]);
+    }
+#endif // #ifdef DEBUG_PLONK
+
+    // permute circuit.H_gen according to the wire permutation
```

Comment:
`circuit.` no longer used AFAICT.

Reply:
Correct. Removed reference to `circuit.`.